### DISCUSSION TITLE
Does the config validator not like customEnvVariables?

### DISCUSSION THREAD

**Discussion by HenrikPoulsen** [renovatebot/renovate#10066] [Request Help] [answered]
On my self-hosted renovate I run the config validator whenever I change the config.js and I noticed that now when I add custom environment variables the validation starts to fail.
Is this expected and I should not use the config validator or is this a bug I should report?
```
ERROR: config.js contains errors
       "errors": [
         {
           "topic": "Configuration Error",
           "message": "Invalid configuration option: customEnvVariables.NPM_CONFIG_REGISTRY"
         },
         {
           "topic": "Configuration Error",
           "message": "Invalid configuration option: customEnvVariables.PIP_INDEX_URL"
         },
         {
           "topic": "Configuration Error",
           "message": "Invalid configuration option: customEnvVariables.YARN_REGISTRY"
         }
       ]

```

**Comment by viceice** [comment]
which version of renovate are you using for validation? It looks like you're using an old version.

**Reply by HenrikPoulsen** [reply]
I am using 25.22.0 which was released yesterday, so not that old

**Reply by viceice** [reply]
ok, how do you run the validator? use `LOG_LEVEL=debug`. show your sanitized `config.js` please.

**Reply by HenrikPoulsen** [reply]
I'm using the renovate npm package so currently I'm running it with:
`node node_modules/renovate/dist/config-validator.js`
And here's my config.js (with some obsfuscated things):
```
module.exports = {
  endpoint: 'someUrl',
  platform: 'github',
  autodiscover: false,
  prCommitsPerRunLimit: 50,
  gitAuthor: "Some User <someUser@somedomain.com>",
  onboardingConfig: {
    extends: [
      ":separateMajorReleases",
      ":combinePatchMinorReleases",
      ":ignoreUnstable",
      ":prImmediately",
      ":semanticPrefixFixDepsChoreOthers",
      "schedule:nonOfficeHours",
      ":automergeDisabled",
      ":ignoreModulesAndTests",
      ":maintainLockFilesDisabled",
      ":autodetectPinVersions",
      ":prHourlyLimit2",
      ":prConcurrentLimit20",
      "group:monorepos",
      "group:recommended",
      "helpers:disableTypesNodeMajor"
    ]
  },
  customEnvVariables: {
    PIP_INDEX_URL: "someUrl",
    NPM_CONFIG_REGISTRY: "someUrl",
    YARN_REGISTRY: "someUrl"
  }
};
```

**Reply by viceice** [reply]
ok, can reproduce it, will debug it now

**Reply by viceice** [reply]
@rarkins It's failing here:
https://github.com/renovatebot/renovate/blob/b8e36daa87e4275caa5b66bffa6fe3f4d5d8a77e/lib/config/validation.ts#L196-L200

because the env key (eg `PIP_INDEX_URL` ) is obviously not in `optionTypes`

**Reply by viceice** [reply]
Looks like we need `freeChoice` option, sending a pr

https://github.com/renovatebot/renovate/blob/b8e36daa87e4275caa5b66bffa6fe3f4d5d8a77e/lib/config/validation.ts#L512-L523

**Reply by HenrikPoulsen** [reply]
I noticed your PR #10070 and tested that out locally. Looks like it is happy with my config with those changes.
Thanks for helping out with this :) Looking forward to getting my hands on it.

**Comment by viceice** [comment, marked as answer]
Fixed in [`25.26.1`](https://github.com/renovatebot/renovate/releases/tag/25.26.1) 🎉

**Reply by HenrikPoulsen** [reply]
Awesome. Thanks for the quick resolution :)
